compiler/damlc/lib/DA/Cli/Damlc/DependencyDb.hs

```diff
@@ -246,24 +279,143 @@ validatePkgId pkgId = do
         fail $ "Invalid package ID dependency in daml.yaml: " <> pkgId
     pure $ LF.PackageId $ T.pack pkgId
 
+-- Package resolution
+---------------------
+
+data FullPkgName = FullPkgName
+    { pkgName :: LF.PackageName
+    , pkgVersion :: LF.PackageVersion
+    } deriving (Eq, Ord, Show)
+
+data LockFile = LockFile
+    { dependencies :: [DependencyInfo]
+    } deriving Generic
+instance Aeson.FromJSON LockFile
+instance Aeson.ToJSON LockFile
+data DependencyInfo = DependencyInfo
+    { name :: String
+    , version :: String
+    , pkgId :: String
+    } deriving Generic
+instance Aeson.FromJSON DependencyInfo
+instance Aeson.ToJSON DependencyInfo
+
+resolvePkgs :: NormalizedFilePath -> Options -> [FullPkgName] -> IO (M.Map FullPkgName LF.PackageId)
+resolvePkgs projRoot opts pkgs
+    | null pkgs = pure M.empty
+    | otherwise = do
+        mbRes <- resolvePkgsWithLockFile lockFp pkgs
+        resOrErr <- case mbRes of
+          Nothing -> resolvePkgsWithLedger depsDir pkgs
+          Just res -> pure $ Right res
+        case resOrErr of
+            Left missing ->
+                fail $
+                unlines $
+                "Unable to resolve the following packages in daml.yaml:" :
+                [ (T.unpack $ LF.unPackageName pkgName) <> "-" <>
+                (T.unpack $ LF.unPackageVersion pkgVersion)
+                | FullPkgName {pkgName, pkgVersion} <- missing
+                ]
+            Right result -> do
+              writeLockFile lockFile result
+              pure result
+  where
+    depsDir = dependenciesDir opts projRoot
+    lockFp = fromNormalizedFilePath projRoot </> lockFile
+
+writeLockFile :: FilePath -> M.Map FullPkgName LF.PackageId -> IO ()
+writeLockFile lockFp resolvedPkgs = do
+    Yaml.encodeFile lockFp $
+        LockFile
+            [ DependencyInfo
+                { name = T.unpack $ LF.unPackageName pkgName
+                , version = T.unpack $ LF.unPackageVersion pkgVersion
+                , pkgId = T.unpack $ LF.unPackageId pkgId
+                }
+            | (FullPkgName {pkgName, pkgVersion}, pkgId) <- M.toList resolvedPkgs
+            ]
+
+-- | Read the package lock file and resolve the given packages. Returns a list of unresolved
```

Comment:
Comment on semantics would be useful:
1. Not exists -> Nothing
2. Exists but invalid format -> Error

Related question, what happens on an SDK upgrade with an incompatible format change? I guess ideally we don’t want to make those so it seems fine to break this but maybe worth documenting.

Reply:
This will fail when an incompatible format change is encountered. The error message will however indicate the user to delete the lock file and re-run. The next invocation will write out a properly formatted lock file.

I added a comment for the exact semantics.